cpp/src/hash/fixed_hash_set.cuh

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <utilities/device_atomics.cuh>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <hash/helper_functions.cuh>
+
+namespace cudf {
+
+/*
+ *  fixed set on a device
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class device_fixed_hash_set {
+public:
+  device_fixed_hash_set(size_type hash_size, const size_type *hash_begin, const Element *hash_data)
+    : hash_size{hash_size}, hash_begin{hash_begin}, hash_data{hash_data}, hasher(), equals() {
+  }
+
+  bool __device__ contains(Element e) const {
+    size_type loc = hasher(e) % (2 * hash_size);
+
+    for (size_type i = hash_begin[loc] ; i < hash_begin[loc+1] ; ++i) {
+      if (equals(hash_data[i],e))
+        return true;
+    }
+
+    return false;
+  }
+
+private:
+  Hasher           hasher;
+  Equality         equals;
+  size_type        hash_size;
+  const size_type *hash_begin;
+  const Element   *hash_data;
+};
+
+/*
+ * Fixed size set on a device.  
+ *
+ */
+template <typename Element, typename Hasher = default_hash<Element>,
+          typename Equality = equal_to<Element>>
+class fixed_hash_set {
+public:
+  /**---------------------------------------------------------------------------*
+   * @brief Factory to construct a new fixed_device_set
+   *---------------------------------------------------------------------------**/
+  static fixed_hash_set<Element> create(column_view const& col,
+                                        cudaStream_t stream) {
+
+    auto d_column = column_device_view::create(col);
+    auto d_col = *d_column;
+
+    rmm::device_vector<size_type> hash_bins_start(2 * d_col.size() + 1, size_type{0});
+    rmm::device_vector<size_type> hash_bins_end(2 * d_col.size() + 1, size_type{0});
+    rmm::device_vector<Element>   hash_data(d_col.size());
+
+    Hasher hasher;
+    size_type *d_hash_bins_start = thrust::raw_pointer_cast(hash_bins_start.data());
+    size_type *d_hash_bins_end = thrust::raw_pointer_cast(hash_bins_end.data());
+    Element *d_hash_data = thrust::raw_pointer_cast(hash_data.data());
+
+    thrust::for_each(rmm::exec_policy(stream)->on(stream),
```

Comment:
This is basically just a histogram of the hash of each key mod the capacity, yes? I'm wondering if you could use a CUB histogram instead.

Reply:
Perhaps.  Since we're going to look at reworking this a different way is it worth worrying about this?

To use the CUB histogram, I think I would need to write a transform iterator that computed the hash.  I'm not sure how I would handle the invalid entries.  Seems like a lot of work for something that ultimately we'll throw away.